[Bug]: Child not rejected when proposed parent fails governance vote

### Problem encountered

```
context = {'AssetID': 'af24231fa0ce43fe6b0a689d9c1ddcd17538d817fefbd7da0fa24b2ca0abd05f', 'app_state.network_parameters': {'bloc...30d5a5ee79dbe4781321f69479141bf7c22216d504e8ce60e9f132fa0de77af2', '0x55Bb84D3FA3A1D8Bd86643F66eaCdfde812fAce5')], ...}
proposer = 'F287A5E197'

    @pytest.mark.successor_markets
    @pytest.mark.markets
    @pytest.mark.full
    def test_parent_rejected_before_child_closes(context, proposer):
        """
        0081-SUCM-027 - Parent market is rejected and proposal closed down before a successor market reaches vote closing/enacted
        """
        # Make a parent which will be declined before the child leaves opening auction
        parent = Market(context, proposer=proposer, closeTimeOffset=9, enactTimeOffset=10)
        parent.propose(proposer, state="Open")
    
        # Propose a child and take it straight out of opening auction
        child = Market(
            context,
            proposer=proposer,
            parent_market_id=parent.id,
            insurance_pool_fraction=0.5,
            submit_lp=False,
            closeTimeOffset=11,
            enactTimeOffset=12,
        )
        child.propose(proposer, state="Open")
    
        marketsTrading.wait_for_market_state(parent.id, expected_state="STATE_REJECTED", timeout=15)
        marketsTrading.wait_for_market_state(child.id, expected_state="STATE_REJECTED", timeout=15)
    
        governMod.wait_for_proposal_state(parent.id, "STATE_DECLINED", wait=15)
>       governMod.wait_for_proposal_state(child.id, "STATE_FAILED", wait=15)

tests/governance/test_markets.py:1031: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

prop_id = '7133df49af057c36640451dfa6d5e674728b26b39ec72158c1eb0f9b891b91f0'
state = 'STATE_FAILED', wait = 15

    def wait_for_proposal_state(prop_id, state="STATE_OPEN", wait=5):
        logging.info(f"looking for proposal {prop_id} in state {state}")
        prop = GetProposalByID(prop_id)
        c = 0
        now = time.time()
        while time.time() <= now + wait:
            if prop["data"]["proposal"]["state"] == state:
                break
            c += 1
            if c % 5 == 0:
                logging.info(f"Did not find expected proposal state, got {prop['data']['proposal']['state']} {prop_id=}")
            time.sleep(0.2)
            prop = GetProposalByID(prop_id)
        logging.info(f"Found proposal state: {prop['data']['proposal']['state']}")
>       assert prop["data"]["proposal"]["state"] == state
E       AssertionError: assert 'STATE_DECLINED' == 'STATE_FAILED'
E         - STATE_FAILED
E         + STATE_DECLINED

modules/govMod/governMod.py:3500: AssertionError
```

### Observed behaviour

In this case, child fails because nobody voted for it (but I can amend the test and the child would actually pass)

### Expected behaviour

Child fails governance when the proposed parent does

### Steps to reproduce

```bash
1. pytest -k test_parent_rejected_before_child_closes
```


### Software version

79cf0344b3896f366220e97c20aa6e2bf2f5eb50

### Failing test

test_parent_rejected_before_child_closes 

### Jenkins run

https://jenkins.ops.vega.xyz/job/common/job/system-tests-nightly/1023/testReport/junit/tests.governance/test_markets/Call_tests___full_g_n___test_parent_rejected_before_child_closes/

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -74,6 +74,7 @@
 - [8874](https://github.com/vegaprotocol/vega/issues/8874) - Database migration can fail when rolling back and migrating up again.
 - [8855](https://github.com/vegaprotocol/vega/issues/8855) - Preserve reference to parent market when restoring checkpoint data
 - [8909](https://github.com/vegaprotocol/vega/issues/8909) - initialise id generator for all branches of market state update
+- [8908](https://github.com/vegaprotocol/vega/issues/8908) - A rejected parent market should result in all successors getting rejected, too.
 
 ## 0.72.1
 

--- a/core/execution/engine.go
+++ b/core/execution/engine.go
@@ -263,6 +263,7 @@ func (e *Engine) RejectMarket(ctx context.Context, marketID string) error {
 		e.log.Debug("reject market", logging.MarketID(marketID))
 	}
 
+	_, isFuture := e.futureMarkets[marketID]
 	if _, ok := e.allMarkets[marketID]; !ok {
 		return ErrMarketDoesNotExist
 	}
@@ -275,7 +276,7 @@ func (e *Engine) RejectMarket(ctx context.Context, marketID string) error {
 	e.broker.Send(events.NewMarketDataEvent(ctx, mkt.GetMarketData()))
 	e.removeMarket(marketID)
 
-	if _, ok := e.futureMarkets[marketID]; !ok {
+	if !isFuture {
 		return nil
 	}
 

--- a/core/governance/engine.go
+++ b/core/governance/engine.go
@@ -340,9 +340,10 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) ([]*ToEnact, []*VoteCl
 	now := t.Unix()
 
 	for _, proposal := range e.activeProposals {
+		// check if the market for successor proposals still exists, if not, reject the proposal
 		if nm := proposal.Terms.GetNewMarket(); nm != nil && nm.Successor() != nil {
 			if _, err := e.markets.GetMarketState(proposal.ID); err != nil {
-				e.rejectProposal(ctx, proposal.Proposal, types.ProposalErrorInvalidSuccessorMarket, ErrInvalidSuccessor)
+				proposal.RejectWithErr(types.ProposalErrorInvalidSuccessorMarket, ErrInvalidSuccessor)
 				e.broker.Send(events.NewProposalEvent(ctx, *proposal.Proposal))
 				toBeRemoved = append(toBeRemoved, proposal.ID)
 				continue

--- a/core/governance/engine.go
+++ b/core/governance/engine.go
@@ -49,6 +49,7 @@ var (
 	ErrSpotsNotEnabled                           = errors.New("spot trading not enabled")
 	ErrParentMarketDoesNotExist                  = errors.New("market to succeed does not exist")
 	ErrParentMarketAlreadySucceeded              = errors.New("the market was already succeeded by a prior proposal")
+	ErrInvalidSuccessor                          = errors.New("the successor is no longer valid")
 )
 
 //go:generate go run github.com/golang/mock/mockgen -destination mocks/mocks.go -package mocks code.vegaprotocol.io/vega/core/governance Markets,StakingAccounts,Assets,TimeService,Witness,NetParams,Banking
@@ -339,22 +340,20 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) ([]*ToEnact, []*VoteCl
 	now := t.Unix()
 
 	for _, proposal := range e.activeProposals {
+		if nm := proposal.Terms.GetNewMarket(); nm != nil && nm.Successor() != nil {
+			if _, err := e.markets.GetMarketState(proposal.ID); err != nil {
+				e.rejectProposal(ctx, proposal.Proposal, types.ProposalErrorInvalidSuccessorMarket, ErrInvalidSuccessor)
+				e.broker.Send(events.NewProposalEvent(ctx, *proposal.Proposal))
+				toBeRemoved = append(toBeRemoved, proposal.ID)
+				continue
+			}
+		}
 		// do not check parent market, the market was either rejected when the parent was succeeded
 		// or, if the parent market state is gone (ie succession window has expired), the proposal simply
 		// loses its parent market reference
 		if proposal.ShouldClose(now) {
 			e.closeProposal(ctx, proposal)
 			voteClosed = append(voteClosed, e.preVoteClosedProposal(proposal))
-		} else if nm := proposal.Terms.GetNewMarket(); nm != nil && nm.Successor() != nil && proposal.IsOpen() {
-			// if the proposal is for a successor market, check if the successor wasn't rejected because another proposal
-			// has since been enacted and left opening auction
-			if _, err := e.markets.GetMarketState(proposal.ID); err != nil {
-				// parent has already been succeeded, the proposal should be rejected
-				proposal.Reject(types.ProposalErrorInvalidSuccessorMarket)
-				e.broker.Send(events.NewProposalEvent(ctx, *proposal.Proposal))
-				toBeRemoved = append(toBeRemoved, proposal.ID)
-				continue
-			}
 		}
 
 		if !proposal.IsOpen() && !proposal.IsPassed() {

